### DISCUSSION TITLE
Feature Request: Allow static import from URL

### DISCUSSION THREAD

**Discussion by TheThirdRace** [vercel/next.js#30063] [Ideas]
### Describe the feature you'd like to request

I would like to be able to use the image `import` feature for markdown content.

I need a way to get automatically the image's `width` and `height`, and the `data-uri` for the blur placeholder.

It would be such a huge time saver.

** Note: **
Please, don't move this to a discussion. It's a graveyard there and the feature request is 100% valid.

Using markdown content is a popular choice and is unfortunately not covered by the `import` method.

### Describe the solution you'd like

Add back the original solution for `static` images.

In the `beta`, static images were using a query parameter in the URL . See PR [#24993](https://github.com/vercel/next.js/pull/24993)

Eventually, that solution was replaced by the `import` statement. See PR [https://github.com/vercel/next.js/pull/25909/files](https://github.com/vercel/next.js/pull/25909/files)

Unfortunately, while `import` is a much nicer solution when dealing with React components, it's also incompatible with markdown content.

Adding back the original query parameter solution would solve the problem for markdown content.

The good news is the code already exists. The links to the 2 PR show everything that needs to be done to reinstate the solution.

### Describe alternatives you've considered

Currently, the only way to use the `import` method for markdown would be:

1. `import` all images used in your markdown content in a Map
2. Put the associated Map key in your front-matter content
3. In `getStaticProps`, use the front-matter key on the Map to select the right information to pass to the page component

Unfortunately, this approach has some very bad downsides...

First, it doesn't scale well if you have multiple images in a page. How do you pass which image is which in markdown or front-matter?

Second, the amount of work to achieve this is very bothersome to say the least. You'd have to manually manage a Map of all your images in your content, which can easily go in the thousands very quickly. Then you have to put the associative logic for the Map in all pages that read markdown content. This is prone to a lot of mistakes...

It would be much faster to simply put a query parameter in the URL directly in markdown. And if I don't want to do this in the content itself to keep it "pure", I can easily add the query parameter in the component that translate images from markdown to a React component.

**Comment by TheThirdRace** [comment]
@timneutkens 
Please, could you elaborate why this was moved to a discussion?

It's the second time this feature request is moved to a discussion. Which is a way to put it in the graveyard...

I'm just trying to understand why this is such a taboo to want to use static images for markdown content?

If I could understand why you're making this decision, I would drop it instantly. But I haven't got any feedback for the old topic either :(

So could you please elaborate on the reasons why this feature request is tossed aside? If only to put my mind at ease...

**Reply by timneutkens** [reply]
Please see the template for feature requests: https://github.com/vercel/next.js/issues/new?assignees=&labels=template%3A+story&template=3.feature_request.yml

The second line of the template says:
> Feature requests will be converted to the GitHub Discussions "Ideas" section.

All feature requests are converted to GitHub discussions to ensure they can be viewed and upvoted by the community. If there are enough upvotes we review them and check if they're relevant to be put on the roadmap for our team or if someone can implement it themselves.

**Reply by TheThirdRace** [reply]
I see, it definitely explains what I missed.

Sorry for duplicating the Feature Request then. I'll still keep this one opened since it's more detailed than the previous one.

Thanks for the explanation and have a nice day!

**Comment by pcjmfranken** [comment]
I'm going to be honest with you, besides something to do with importing images using `next/image` in markdown, I haven't the slightest clue as to what you are hoping to achieve.

Can you link to or post some example code of what you are doing now, and some that shows how you would ideally like it to work?

Also stop abusing the issues system for attention. You say the code is there already... why not a PR?

**Reply by TheThirdRace** [reply]
@pcjmfranken 

The reasons to use `import` to get a static image are:

- You get the `width` & `height` directly from the image file itself
- You get a blur placeholder
- The cache on the generated image is `public,max-age=31536000,immutable` which means you don't use any network request for revalidating (faster, less bandwidth, works offline, etc.)

When using markdown, you can only link to your image like so:

```markdown
![Image alt text here](/path/to/image.png)
```

When using a package to render markdown, you get the possibility to overwrite the components. This way, you can tell the markdown renderer to use a custom component, lets call `<NextImage>`, to render an `<img>`. 

The props you receive in `<NextImage>` from the markdown renderer in this case would be the image's `alt` and `src`. That's normal because that's the only information markdown actually has.

Because you only get an `src`, images don't benefit from getting the `width`, the `height` and the blur placeholder automatically. They also revert to the default caching strategy (which is inferior to the immutable strategy).

Before the `import` method was introduced in NextJs 11.0, the beta version had a method to use a query parameter in the `src`.

For example, you could use an `src` like this: `/path/to/image.png?s=1`

NextJs backend would automatically pick the `s` parameter and apply the `static` importing. Thus, you would get the `width`, the `height`, the blur placeholder and the better caching strategy.

That code was introduced in the following PR: https://github.com/vercel/next.js/pull/24993

Eventually, that method was replaced by the `import` method that is much easier for React components.

The keyword is "replaced", not a new method added, they removed the previous method because it was deemed less elegant. (which is true in a sense)

Unfortunately, the beta method of using a query  parameter would have solved all markdown problems:

```markdown
![Image alt text here](/path/to/image.png?s=1)
```

Or you simply adjust a `<NextImage>` component to append `?s=1` to the `src`.

**So, as a feature request, I would like NextJs to reintroduce the query parameter method.**

I hope this clears up a bit more what I'm asking for.

**Reply by TheThirdRace** [reply]
@pcjmfranken 
I created a second reply to address the final part of your message.

> Also stop abusing the issues system for attention.

I'm sorry, but this was uncalled for.

I created a feature request on August 3rd 2021, more than 2 months ago. 

It was pretty much immediately moved to a discussion ~and ignored by NextJs team~. I did ask if I could improve the feature request, but never got any answer.

On October 19th 2021, **after more than 2 months**, I decided to recreate the feature request as this is a very valid use case and I still need it.

It took 2 hours to be moved again to a discussion, ~forever to be ignored by NextJs team...~

~All the while, plenty of feature requests are left in the `Issues` system for months, if not years.~

~I just don't get why NextJs is so prompt in brushing this subject aside. And I unfortunately don't get any answers either.~

So no, I'm not abusing the `Issues` system to get attention. Please refrain from insinuating this in the future.

Edit: I got an answer from NextJs team, I adjusted this post accordingly.

Long story short, I made a mistake and I'll gladly own it.

I'm leaving this post up because my point still stands. I wasn't abusing the `Issues` system, but given it took 2 months, 2 DMs and a couple Discord visits to get an answer to my confusion, it's understandable I was a bit frustrated...

In any case, I got my answer now and we'll see if it gets implemented down the line.
